### PR TITLE
Enhance recombination algorithm to support 4x4 matrices (#4145)

### PR DESCRIPTION
**Pull request by ton11797**
This pull request enhances the recombination algorithm to support 4x4 matrices. 

**Changes:**
- Modified the recombination function to handle 4x4 matrices.
- Added test cases to verify the correctness of the implementation for 4x4 matrices.

**Related Issues:**
- Issue #4145 

This is my first contribution to public project, so any feedback or suggestions for improvement would be greatly appreciated.

### CODE REVIEW

File: lib/operation.js [hunk 1]
```diff
@@ -787,24 +787,28 @@ function linear (a, b) {
...
-    inputMatrix[2][0], inputMatrix[2][1], inputMatrix[2][2]
-  ].map(Number);
+
+  for (const row of inputMatrix) {
+    if (row.length !== length) {
+      throw new Error('Invalid recombination matrix');
+    }
+  }
+
+  this.options.recombMatrix = inputMatrix.flat().map(Number);
```
Comment (lovell) [inline lib/operation.js:810]: Excellent use of `flat()` here, I suspect it wasn't fully-available when this feature was first added. Perhaps this also provides a simpler way to validate the matrix length, i.e. must be either 9 or 16?

File: lib/index.d.ts [hunk 2]
```diff
@@ -1730,6 +1730,7 @@ declare namespace sharp {
 
     type Matrix2x2 = [[number, number], [number, number]];
```
Comment (lovell) [inline lib/index.d.ts:1731]: Not sure `Matrix2x2` is used anywhere, I suspect this was inherited via DefinitelyTyped and could be removed.

**Review by lovell — COMMENTED** [review]
Fantastic, thank you very much for working on this, I've left a couple of questions inline.

File: lib/index.d.ts (hunk 2 above)
Comment (ton11797) [reply lib/index.d.ts:1731]: this type are used at affine() //2x2 affine transformation matrix

File: lib/operation.js (hunk 1 above)
Comment (ton11797) [reply lib/operation.js]: fixed

File: src/pipeline.cc [hunk 3]
```diff
@@ -1613,10 +1613,16 @@ Napi::Value pipeline(const Napi::CallbackInfo& info) {
     }
   }
   if (options.Has("recombMatrix")) {
-    baton->recombMatrix = std::unique_ptr<double[]>(new double[9]);
+    baton->recombMatrixSize = sharp::AttrAsInt32(options, "recombMatrixSize");
```
[the remark this answers is not in the document]
Comment (ton11797) [reply src/pipeline.cc]: fixed

**Comment by lovell** [comment]
Thanks again for adding this feature, I made a couple of small changes and it's now on the `main` branch as commit https://github.com/lovell/sharp/commit/60c5c5083dd12fae04db2640bb5c40564bb0728d for inclusion in the v0.33.5 release.

**Pull request closed without merging**